multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/SCMBinder.java

```diff
@@ -0,0 +1,134 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.multibranch;
+
+import com.thoughtworks.xstream.converters.Converter;
+import groovy.lang.GroovyShell;
+import hudson.Extension;
+import hudson.model.Action;
+import hudson.model.ItemGroup;
+import hudson.model.Queue;
+import hudson.model.TaskListener;
+import hudson.scm.SCM;
+import hudson.util.DescribableList;
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.WeakHashMap;
+import jenkins.branch.Branch;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMHeadObserver;
+import jenkins.scm.api.SCMRevision;
+import jenkins.scm.api.SCMSource;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowExecution;
+import org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition;
+import org.jenkinsci.plugins.workflow.cps.GroovyShellDecorator;
+import org.jenkinsci.plugins.workflow.flow.FlowDefinition;
+import org.jenkinsci.plugins.workflow.flow.FlowDefinitionDescriptor;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.pickles.Pickle;
+import org.jenkinsci.plugins.workflow.support.pickles.SingleTypedPickleFactory;
+import org.jenkinsci.plugins.workflow.support.pickles.XStreamPickle;
+
+/**
+ * Adds an {@code scm} global variable to the script.
+ * This makes it possible to run {@code checkout scm} to get your project sources in the right branch.
+ */
+class SCMBinder extends FlowDefinition {
+
+    private static final Map<CpsFlowExecution,SCM> scms = Collections.synchronizedMap(new WeakHashMap<CpsFlowExecution,SCM>());
+
+    @Override public FlowExecution create(FlowExecutionOwner handle, TaskListener listener, List<? extends Action> actions) throws Exception {
+        Queue.Executable exec = handle.getExecutable();
+        if (!(exec instanceof WorkflowRun)) {
+            throw new IllegalStateException("inappropriate context");
+        }
+        WorkflowJob job = ((WorkflowRun) exec).getParent();
+        BranchJobProperty property = job.getProperty(BranchJobProperty.class);
+        if (property == null) {
+            throw new IllegalStateException("inappropriate context");
+        }
+        Branch branch = property.getBranch();
+        ItemGroup<?> parent = job.getParent();
+        if (!(parent instanceof WorkflowMultiBranchProject)) {
+            throw new IllegalStateException("inappropriate context");
```

Comment:
It is an `IllegalStateException` because the user should never see them unless I messed up somehow (or they did something weird with the script console, etc.).